Python/chapter01/1.8 - Zero Matrix/miguel_1.8_sol.py

```diff
@@ -0,0 +1,135 @@
+"""
+Python version 3.7.0
+1.8 - Zero Matrix
+Write an algorithm such that if an element in an MxN matrix is 0, its entire row and column are set to 0
+"""
+import unittest
+from typing import List
+
+
+def zero_matrix(matrix: List[List[int]]) -> List[List[int]]:
+    """
+    zero_matrix will take in an MxN matrix and when a 0 element is found, the whole column and row
+    will be set to 0.
+    Runtime: O(M * N)
+    Space Complexity: O(min(M, N))
+    :param matrix: an MxN matrix.  M is the number of rows, N is the number of columns
+    :return: a zero-modified matrix
+    """
+    M = len(matrix)
+    N = len(matrix[0])
+
+    zeroed_rows = set()
+    zeroed_cols = set()
+
+    for row in range(M):
+        for col, num in enumerate(matrix[row]):
+            if num != 0 or row in zeroed_rows or col in zeroed_cols:
+                continue
+            # otherwise, set row to 0 by looping through columns of current row
+            for i in range(N):
+                matrix[row][i] = 0
+            # set column to 0 by looping through rows of current column
+            for j in range(M):
+                matrix[j][col] = 0
+            # update zeroed row and col sets
+            zeroed_cols.add(col)
+            zeroed_rows.add(row)
```

Comment:
just a small optimization: you could add a `break` here to stop considering the rest of the elements of the row since you know that `row in zeroed_rows` will be `True` in the rest of the elements.